### PR TITLE
Simplification pipeline

### PR DESCRIPTION
**Pull request by martinfleis**
This kind of contains everything but will clean and squash it once we'll close this part.

### CODE REVIEW

**Comment by anastassiavybornova** [comment]
Dropping some comments on the current `nx_gx` function (after some Wuhan exploration). Suggested (unfinished!!) rephrasing:

Drop all but highest hierarchy. 

If there are **three or more** edges of the highest hierarchy, use roundabout solution. (this does by construction guarantee that we won't disconnect anything right?)

If there are **two** edges of highest hierarchy: 

a. If there are no further nodes to connect: check whether the 2 edges are now disconnected; if so, reconnect (shortest line if inside artifact, else skeleton)
  
b. If there ARE further nodes to connect (from prime or lower-hierarchy intersections): _[our initial suggestion was connect them all to one point ideally, but this creates spaghetti]_ --> this one needs to be elaborated!

If there is only **one** edge of highest hierarchy:

...? (probably same problem as for 2-edge above in case of interstitial nodes to be connected)

Good example to start looking at this (Wuhan): 5CEES

image


***

Re "Connection point should ideally be an existing nearest node with degree 4 or above": counterexample: Wuhan 5CEES

image

**Comment by anastassiavybornova** [comment]
In other & better news, I checked all cases where one of the other 2 functions (`n1_g1_identical` [maybe rename to n1_g1? since 1 is 1 by default] and `nx_gx_identical`) would apply for Wuhan, and it all seems to work, but would like to doublecheck with you @martinfleis whether we're fine where dual carriageways are concerned (eg 2EE, 2CC, 3CC) and whether I reasoned correctly for 4EEE

**Comment by anastassiavybornova** [comment]
* re n1g1: check elongation and conditionally apply dead end solution (possibly hausdorff point approximation by cutting linestring in 2; convex hull thingy)
* possible update to nxgx identical: proportionally to area and perimeter create some elongation rule and skeletonize (when there are 2 or 3 nodes we can always do it)
* centroid or direct connection - depending on which one is shorter
* skeletonize!!

**Comment by anastassiavybornova** [comment]
**Strategy for 2-artifact clusters (first draft):**
* Get linestring geometry of interline
* Find which coin the interline belongs to
* Find whether the coin ends on the artifact cluster
* (C) if no: continuous case >> treat artifacts separately
* (E) if yes, and if coin is not entirely covered by artifact cluster >> E - drop both interedge but keep internodes (possibly: first of the 2?); then treat as single artifact
* (S) if yes, and if coin is entirely covered by artifact cluster >> S - drop both interedge and its internodes; then treat as single artifact

Caveats:
* for treating a cluster as single, we need to recompute typology
* dropping edges == modifying street network edges and hence everything else (incl. coins)! (possibly re-run the entire thing?)

**Comment by martinfleis** [comment]
> (C) if no: continuous case >> treat artifacts separately

this may be tricky as if we're connecting both to C, they shall be ideally connected to the same node on C to have only one intersection as a result. So treating them completely separately may not be possible. (_midnight thoughts_)

**Comment by anastassiavybornova** [comment]
**Types of nxgx issues in Liège**

after checking *all* artifact solutions of type nXY where X=mC and Y in {pE, rS}

***

image
(4CCSS, 5CCSS, 7CCSS, 7CCCSS, 4CSS) !! dead ends within artifact introduced

***

image
(5CES, 5CEES, 6CEES, 7CES, 5CCES) !! not fully connected

***

image
(4CEE, 4CCEE) !! new artifact introduced

***

image
(8CS) !! primes not snapped

**Comment by anastassiavybornova** [comment]
**identifying contiguous artifacts** ...and the code goes something like this:
```
rook = graph.Graph.build_contiguity(artifacts, rook=True)
artifacts['comp'] = rook.component_labels
counts = artifacts['comp'].value_counts()
artifacts = artifacts.loc[artifacts['comp'].isin(counts[counts==2].index)]
```

File: core/algorithms/simplify.py
Comment (jGaboardi) [inline core/algorithms/simplify.py:1]: 👀

**Comment by martinfleis** [comment]
Hey, sorry that this is so massive but it just contains everything. It was much easier to do it this way than split, it was veeeery iterative work.

the `simplification_pipeline` notebook contains the current top-level functionality for simplification. It runs on all cases apart from Wuhan, where cluster simplification fails and SLC, where it runs but simplifies a tiny fraction due to some issues with artifact detection.

I would probably merge it and iterate further in subsequent PRs. No idea if this is review-able, probably not really.

The code is still a bit messy with plenty of opportunities to streamline and clean it.

**Comment by anastassiavybornova** [comment]
@jGaboardi 
what i  wanted to suggest here re notebook renaming - so that it becomes clearer - 
* the current logic distinguishes between "singles", "doubles", and "multiples" which maps to the notebooks "typology", "typology-doubles", and "typology-multiples" >> suggestion to rename typology as "typology-single"
* as martin just said, the `triangle` and `rectangle` ones are superseded but still suggestion to rename into "typology-triangles" and "typology-rectangles"

**Comment by jGaboardi** [comment]
Ahhhh. OK, got it. So:

* `typology.ipynb` --> `typology-single.ipynb`
* `triangles.ipynb` --> `typology-triangles.ipynb`
* `rectangles.ipynb` --> `typology-rectangles.ipynb`

Right?

**Comment by anastassiavybornova** [comment]
@jGaboardi , yes, and after renaming, you can ignore the `typology-triangles` and `typology-rectangles` ones :)

**Comment by jGaboardi** [comment]
@anastassiavybornova 

Do we think `typology-single` or `typology-singles`?

**Comment by anastassiavybornova** [comment]
singles. to be consistents.

**Review by jGaboardi — APPROVED** [review]

File: notebooks/typology.ipynb
[the remark this answers is not in the document]
Comment (martinfleis) [reply notebooks/typology.ipynb:1]: yeah

**Pull request merged**